File: kotlin_codeql/src/main/java/org/example/kotlin_codeql/Security/CWE/CWE-190/ArithmeticWithExtremeValues.kt
// {fact rule=arithmetic-overflow@v1.0 defects=0}
internal object ArithmeticWithExtremeValues {
    @JvmStatic
    fun main(args: Array<String>) {
        run {
            val i = Long.MAX_VALUE
            // BAD: overflow
            val j = i + 1
        }
        run {
            val i = Int.MAX_VALUE
            // GOOD: no overflow
            val j = i.toLong() + 1
        }
    }
}
// {/fact}

File: cpp_codeql/Security/CWE/CWE-190/ArithmeticWithExtremeValues.cpp
#include <iostream>
#include <climits>

namespace {
    int main(int argc, char** argv) {
        int i = INT_MAX;
        // {fact rule=integer-overflow@v1.0 defects=1}
        // BAD: overflow
        int j = i + 1;
        // {/fact}

        // ...

        int l = INT_MAX;
        // {fact rule=integer-overflow@v1.0 defects=0}
        // GOOD: no overflow
        long k = (long)l + 1;
        // {/fact}
    }
}

